### ISSUE TITLE
Searching for citric acids in global search leads to error

### ISSUE THREAD

**Issue by inghylt** [MetabolicAtlas/MetabolicAtlas#765]
**Describe the bug**

When searching for citric acid in quick search, and then going to global search from there, leads to an error 

**Steps to reproduce**

1. Search for citric acid in quick search
2. Press link to global search
3. Note the error being thrown

**Expected behavior**

That going from quick to global search works as when searching for citric acid in just global search, i.e. that only `No matches for citric acid` will be returned and not an error message

**Screenshots**

![image.png](https://images.zenhubusercontent.com/5fc0f648ba2981921764866b/2c1e36fc-1441-4a1a-b945-04ed227e19c9)

**Additional context**

Note that this is only applicable when going from quick search to global search, and not when searching in global search directly

**Comment by mihai-sysbio** [comment]
Here is another problematic link: https://metabolicatlas.org/search?term=Acrylic%20acid

**Comment by inghylt** [comment]
Some initial research:
- the error is thrown if you go from quick search to global, or if you visit the global search with a search term directly in the url. For example, visiting `/search?term=C5-branched dibasic acid metabolism`leads to an error but typing `C5-branched dibasic acid metabolism` and then searching does not
- the url is updated after the result for a typed search is returned, not when you press enter
- As noted above, global search does not seem to find all relevant matches. `C5-branched dibasic acid metabolism` and `Tricarboxylic acid cycle and glyoxylate` does not lead to a match even though it is (part of the name of) a subsystem. `Tricarboxylic acid cycle and glyoxylate ` (note the added space) does work though. They are however found in the quick search. Have not yet found a dysfunctional one that does not have acid in the name

**Comment by inghylt** [comment]
This is interesting. Seems like I can't reproduce the error on localhost and results are actually returned.

On localhost it takes around 7-10 seconds, on metabolicatlas.org around 15-17 seconds. That should not trigger the timeout on metabolicatlas.org right?

Can someone confirm they don't get the problem on localhost either? Try links `localhost/search?term=fatty acids`, `http://localhost/search?term=citric%20acid` `http://localhost/search?term=citric%20acid`and `http://localhost/search?term=C5-branched%20dibasic%20acid`. These paths do create problems on metabolicatlas.org which seem to be deployed with the latest version of develop

**Comment by inghylt** [comment]
"The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout."

Makes sense since we in `docker-compose-remote.yml` have:
neo4j:
    environment:
      - NEO4J_dbms_transaction_timeout=15s

So increasing that should solve it. Does 30s sound like a good idea to you guys? That's our fail_timeout in nginx.conf

**Comment by mihai-sysbio** [comment]
The 15s timeout was quite intentional. Let's try to keep the timeout and create another issue to speed up the search - waiting for 15s for a search doesn't really make for a pleasant user experience. Plus, this is holding up the database in query, so we run the risk of having failed requests because of having too many long requests in parallel.

**Comment by inghylt** [comment]
Ok. Should we try to make some kind temporary fix for this for now, like catching the error and not display "Something went wrong, please try again later or contact us at contact@metabolicatlas.org.". Or is it better to keep things as they are for now to mark that we do have backend problems and not "trick them" to believe that there are no results?

**Comment by nanjiangshu** [comment]
Probably this PR https://github.com/MetabolicAtlas/MetabolicAtlas/pull/880 fixes this issue. When I tested with the branch `fix/quick-search-error-chars `. I don't see the error message.

**Comment by nanjiangshu** [comment]
Aha, now I see the problem. You are right, it is indeed a timeout issue. Maybe we can make the error message more precisely.

**Comment by MalinAhlberg** [comment]
Nice research @inghylt! I agree we should update the error message to something more specific like "This search took too long".

**Comment by inghylt** [comment]
Conclusion from backlog refinement: Try to fix response time through #101 and do not mess with error messages for now

**Comment by inghylt** [comment]
Explore neo4j paral, timecap 8 h

**Comment by inghylt** [comment]
@inghylt and @e0 have explored to see if neo4j parallelization can speed up the neo4j response time and thus fix this bug. The explorations are available on [this branch](https://github.com/MetabolicAtlas/MetabolicAtlas/commits/feat/search-parallelization). It seems as parallelization does speed up individual queries. For example, when running the code related to fetchCompartmentalizedMetabolites in the neo4j browser the response time was cut in half. However, when all the queries related to global search are run together they seem to perform worse than current implementation on develop. Different parallelization configurations and CPU/MEM resources have been tested but no success with that either. An attempt to remove the js paralellization and only rely on neo4j paralellization was made as well as they might not work well together, but this did not seem to speed up things either.

Kind of a bummer that we can't see an improvement. Good thing though is that the knowledge regarding neo4j paralellization has been spread within the team. We will make a summary of current neo4j paralellization knowledge on the metatlas wiki, will post a comment here when that is done.

Regarding the citric acid bug, we suggest we increase NEO4J_dbms_transaction_timeout to a higher number, say 25 sek, and restore it to 15 sek once we have pagination in place. There is no other possible fix available at the moment and it is not like we will go lazy and accept non performative code because of a higher limit

**Comment by e0** [comment]
> Thanks @inghylt and @e0 for the great summary about the parallelization. I think it is a good idea to stop here. Before the pagination will be implemented, I'm wondering if it would be possible temporary set up a notification to the user when it hits TIMEOUT and ask if they want to continue, similar to what we have implemented in the Interaction Parters previously? If it takes a signification amount of time to implement this UX, then I would opt for Ingrid's suggestion, that is, increase `NEO4J_dbms_transaction_timeout ` to a higher number.

That's a good idea. It should be possible to make some changes for this to be possible but I don't know if it's something that is worth putting time in as a temporary fix, especially considering we usually only deploy to production once a sprint.

**Comment by MalinAhlberg** [comment]
Thanks for the summary! Having a section in the wiki about this sounds very useful!
I also agree that it makes sense to stop here, given that we're working on the pagination and improvement of the global search. Increasing the timeout sounds reasonable to me.

**Comment by mihai-sysbio** [comment]
And with all the agreement, I think this issue is resolved and can be closed.

**Comment by inghylt** [comment]
Summary has been written and added to the wiki under section Development/neo4j
